javascript/ql/lib/semmle/javascript/internal/flow_summaries/Strings.qll

```diff
@@ -64,3 +66,33 @@ class StringSplit extends SummarizedCallable {
     output = "ReturnValue.ArrayElement"
   }
 }
+
+/**
+ * A call of form `x.split("#")` or `x.split("?")`.
+ *
+ * These are of special significance when tracking a tainted URL suffix, such as `window.location.href`,
+ * because the first element of the resulting array should not be considered tainted.
+ *
+ * This summary defaults to the same behaviour as the general `.split()` case, but it contains optional steps
+ * and barriers named `tainted-url-suffix` that should be activated when tracking a tainted URL suffix.
+ */
+class StringSplitHashOrQuestionMark extends SummarizedCallable {
+  StringSplitHashOrQuestionMark() { this = "String#split with '#' or '?'" }
+
+  override DataFlow::MethodCallNode getACallSimple() {
+    result.getMethodName() = "split" and
+    result.getNumArgument() = [1, 2] and
+    result.getArgument(0).getStringValue() = ["#", "?"]
+  }
+
+  override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+    preservesValue = false and
+    (
+      input = "Argument[this].OptionalBarrier[tainted-url-suffix]" and
+      output = "ReturnValue.ArrayElement"
```

Comment:
I'm not sure how this only blocks taint on the first element? In my mind `ReturnValue.ArrayElement` should block all array elements.  
The tests clearly shows that it works, I'm just not sure how.

Reply:
It blocks all flow into the `ReturnValue.ArrayElement` output. But the second summary below re-establishes flow into `ArrayElement[1]` via a different data flow path.